relayer/relays/beacon/header/syncer/syncer.go

```diff
@@ -493,55 +523,120 @@ func (s *Syncer) unmarshalBeaconState(slot uint64, data []byte) (state.BeaconSta
+	}
+
+	for {
+		finalizedSlot, attestedSlot, err := s.findValidUpdatePair(slot)
 		if err != nil {
-			slot -= s.setting.SlotsInEpoch
+			if highestSlot < slot || head.Slot < slot {
+				return 0, fmt.Errorf("unable to find valid slot")
+			}
+
+			slot += s.protocol.Settings.SlotsInEpoch
 
 			continue
 		}
 
-		headers = append(headers, header.Slot)
-		headers = append(headers, finalizedHeader.Slot)
+		log.WithFields(log.Fields{"attested": attestedSlot, "finalized": finalizedSlot}).Info("found boundary headers")
+		return attestedSlot, nil
+	}
+}
+
+func (s *Syncer) findValidUpdatePair(slot uint64) (uint64, uint64, error) {
+	finalizedHeader, err := s.Client.GetHeaderBySlot(slot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get finalized slot: %d err: %w", slot, err)
+	}
+
+	attestedSlot := finalizedHeader.Slot + (s.protocol.Settings.SlotsInEpoch * 2)
+	attestedHeader, err := s.Client.GetHeaderBySlot(attestedSlot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get attested slot: %d err: %w", attestedSlot, err)
+	}
+
+	nextHeader, err := s.FindBeaconHeaderWithBlockIncluded(attestedSlot + 1)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get next header: %d err: %w", attestedSlot+1, err)
+	}
+	nextBlock, err := s.Client.GetBeaconBlockBySlot(nextHeader.Slot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get next block: %d err: %w", nextHeader.Slot, err)
+	}
+
+	superMajority, err := s.protocol.SyncCommitteeSuperMajority(nextBlock.Data.Message.Body.SyncAggregate.SyncCommitteeBits)
+	if err != nil {
+		return 0, 0, fmt.Errorf("compute sync committee supermajority: %d err: %w", nextHeader.Slot, err)
+	}
+	if !superMajority {
+		return 0, 0, fmt.Errorf("sync committee at slot not supermajority: %d", nextHeader.Slot)
+	}
+
+	return finalizedHeader.Slot, attestedHeader.Slot, nil
+}
+
+func (s *Syncer) GetLatestPossibleFinalizedUpdate(attestedSlot uint64, boundary uint64) (scale.Update, error) {
+	attestedSlot, err := s.FindLatestAttestedHeadersAtInterval(attestedSlot, boundary)
+	if err != nil {
+		return scale.Update{}, fmt.Errorf("cannot find blocks at boundaries: %w", err)
 	}
 
-	log.WithField("headers_found", headers).Info("found boundary headers")
-	return headers[0], nil
+	return s.GetFinalizedUpdateAtAttestedSlot(attestedSlot, boundary, false)
 }
 
-func (s *Syncer) GetFinalizedUpdateAtAttestedSlot(attestedSlot uint64, lastSyncedFinalizedSlot uint64) (scale.Update, error) {
-	var update scale.Update
+func (s *Syncer) GetFinalizedUpdateWithSyncCommittee(syncCommitteePeriod, lastFinalizedSlot uint64) (scale.Update, error) {
+	slot := (syncCommitteePeriod) * s.protocol.Settings.SlotsInEpoch * s.protocol.Settings.EpochsPerSyncCommitteePeriod
+
+	boundary := (syncCommitteePeriod + 1) * s.protocol.Settings.SlotsInEpoch * s.protocol.Settings.EpochsPerSyncCommitteePeriod
 
-	attestedSlot, err := s.findAttestedAndFinalizedHeadersAtBoundary(attestedSlot, lastSyncedFinalizedSlot)
+	attestedSlot, err := s.FindOldestAttestedHeaderAtInterval(slot, boundary)
```

Comment:
The methods do similar but different things. `FindLatestAttestedHeadersAtInterval ` finds the latest available slot, while `FindOldestAttestedHeaderAtInterval` finds the oldest available slot. I have merged the common code into a method called `findValidUpdatePair` here: https://github.com/Snowfork/snowbridge/pull/1164/files#diff-c589b232c311e67a8690c53fe1886c463a185178fde8ed4aa9dd2def499603b3R578. I think the 2 methods are now small and different enough to warrant 2 functions.

Reply:
Yeah, I understand that. 

Just curious seems we use `FindLatestAttestedHeadersAtInterval` for `syncInterimFinalizedUpdate`  while `FindLatestAttestedHeadersAtInterval` for `SyncCommitteePeriodUpdate`, is that a special consideration for that?

> So can the search `s.FindOldestAttestedHeaderAtInterval(slot, boundary)` be replaced with `s.FindLatestAttestedHeadersAtInterval(boundry, slot)`

Or be more specific as mentioned above is that possible to always search for the latest in a reverse manner?

Reply:
I started typing a response and never sent it. 😅 The reason why these two are different in my opinion is:
- Finalized Update: Ideally want a newer update, so that it can cover a "larger" range of `SLOTS_PER_HISTORICAL_ROOT` (meaning, we want a finalized header as far from the previous finalized header) as we can.
- Finalized Update with Sync Committee: for this case, an earlier update is better since we'd like to update the next sync committee as soon as we can.

Let me know if you think this is kinda unnecessary, and if we should just use the latest header we can (and scrap the "earlier" update for Finalized Update with Sync Committee.